packages/dashboard/src/widget-resize-controller.js

```diff
@@ -0,0 +1,186 @@
+    } else if (this.__resizeHeight < currentElementHeight - rowMinHeight / 2) {
+      // Resized vertically below the half of the current row, decrease rowspan
+      this.__updateResizedItem(this.resizedItem.colspan, Math.max((this.resizedItem.rowspan || 1) - 1, 1));
+    }
+  }
+
+  /** @private */
+  __onResizeEnd() {
+    if (!this.resizedItem) {
+      return;
+    }
+
+    // If the originally resized element is restored to the DOM (as a direct child of the host),
+    // to make sure "track" event gets dispatched, remove it to avoid duplicates
+    if (this.__resizedElement.parentElement === this.host) {
+      this.__resizedElement.remove();
+    }
+
+    const itemWrapper = this.__getItemWrapper(this.resizedItem);
+    itemWrapper.style.removeProperty('--_vaadin-dashboard-widget-resizer-width');
+    itemWrapper.style.removeProperty('--_vaadin-dashboard-widget-resizer-height');
+
+    this.host.$.grid.toggleAttribute('resizing', false);
+
+    // Disconnect the observer for the resized element removal
+    this.__resizedElementRemoveObserver.disconnect();
+    // Cleanup the touchmove listener
+    this.cleanup();
+
+    // Dispatch the resize end event
+    this.host.dispatchEvent(
+      new CustomEvent('dashboard-item-resize-end', {
+        detail: { item: this.resizedItem },
+        cancelable: true,
+      }),
+    );
+    this.resizedItem = null;
+  }
+
+  /** @private */
+  __getElementItem(element) {
+    return element.closest(WRAPPER_LOCAL_NAME).__item;
+  }
+
+  /** @private */
+  __getItemWrapper(item) {
+    return [...this.host.querySelectorAll(WRAPPER_LOCAL_NAME)].find((el) => el.__item === item);
+  }
+
+  /** @private */
+  __updateResizedItem(colspan = 1, rowspan = 1) {
+    if ((this.resizedItem.colspan || 1) === colspan && (this.resizedItem.rowspan || 1) === rowspan) {
+      return;
+    }
+
+    const resizeEvent = new CustomEvent('dashboard-item-drag-resize', {
+      detail: { item: this.resizedItem, colspan, rowspan },
+      cancelable: true,
+    });
+
+    // Dispatch the resize event and resize items if the event is not canceled
+    if (!this.host.dispatchEvent(resizeEvent)) {
+      return;
+    }
+
+    this.resizedItem.colspan = colspan;
+    this.resizedItem.rowspan = rowspan;
+    this.host.items = [...this.host.items];
+    requestAnimationFrame(() => this.__updateWidgetStyles());
+  }
+
+  /** @private */
+  __updateWidgetStyles() {
+    const itemWrapper = this.__getItemWrapper(this.resizedItem);
+    itemWrapper.style.setProperty('--_vaadin-dashboard-widget-resizer-width', `${this.__resizeWidth}px`);
+    itemWrapper.style.setProperty('--_vaadin-dashboard-widget-resizer-height', `${this.__resizeHeight}px`);
+  }
+
+  /** @private */
+  __restoreResizedElement() {
+    if (!this.host.contains(this.__resizedElement)) {
+      this.__resizedElement.style.display = 'none';
+      this.host.appendChild(this.__resizedElement);
+    }
+  }
+
+  cleanup() {
```

Comment:
**nit:** could use hostConnected, hostDisconnected instead

Reply:
Thanks, that's a good tip. Pushed a refactor to use `hostDisconnected` to clean up the listener